exporter/collector/metricsexporter.go

```diff
@@ -156,57 +194,95 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pdata.Metrics) err
 			mes := ilm.Metrics()
 			for k := 0; k < mes.Len(); k++ {
 				metric := mes.At(k)
-				timeSeries = append(timeSeries, me.mapper.metricToTimeSeries(monitoredResource, metricLabels, metric)...)
+				for _, ts := range me.mapper.metricToTimeSeries(monitoredResource, metricLabels, metric) {
+					me.timeSeriesC <- ts
+				}
 
 				// We only send metric descriptors if we're configured *and* we're not sending service timeseries.
 				if me.cfg.MetricConfig.SkipCreateMetricDescriptor || me.cfg.MetricConfig.CreateServiceTimeSeries {
 					continue
 				}
 
 				for _, md := range me.mapper.metricDescriptor(metric) {
-					if md != nil {
-						select {
-						case me.mds <- md:
-						default:
-							// Ignore drops, we'll catch descriptor next time around.
-						}
+					if md == nil {
+						continue
+					}
+					select {
+					case me.metricDescriptorC <- md:
+					default:
+						// Ignore drops, we'll catch descriptor next time around.
 					}
 				}
 			}
 		}
 	}
 
 	// TODO: self observability
+
+	return nil
+}
+
+func (me *MetricsExporter) exportPendingTimeSerieses() {
+	ctx := context.Background()
+
+	var sendSize int
+	if len(me.pendingTimeSerieses) < sendBatchSize {
+		sendSize = len(me.pendingTimeSerieses)
+	} else {
+		sendSize = sendBatchSize
+	}
+
+	var ts []*monitoringpb.TimeSeries
+	ts, me.pendingTimeSerieses = me.pendingTimeSerieses, me.pendingTimeSerieses[sendSize:]
+
+	var err error
 	if me.cfg.MetricConfig.CreateServiceTimeSeries {
-		err := me.createServiceTimeSeries(ctx, timeSeries)
-		recordPointCount(ctx, len(timeSeries), m.DataPointCount()-len(timeSeries), err)
-		return err
+		err = me.createServiceTimeSeries(ctx, ts)
+	} else {
+		err = me.createTimeSeries(ctx, ts)
+	}
+
+	var st string
+	if err == nil {
+		st = "OK"
+	} else if s, ok := status.FromError(err); ok {
+		st = statusCodeToString(s)
+	} else {
+		st = "UNKNOWN"
 	}
-	err := me.createTimeSeries(ctx, timeSeries)
-	recordPointCount(ctx, len(timeSeries), m.DataPointCount()-len(timeSeries), err)
-	return err
+
+	recordPointCountDataPoint(ctx, len(ts), st)
+	me.obs.log.Error("could not export time series to GCM", zap.Error(err))
```

Comment:
do we only want to do this if err != nil?

Reply:
Done